File: sentimental-credit/credit.py
import sys
import math


number = 0

while int(number) <= 0:
    try:
        number = input("Number: ")
    except:
        continue

l = len(number)

if l != 15 and l != 16 and l != 13:
    print("INVALID")
    sys.exit()

cardNum = int(number)
cardNumDigits = int(number)
checkSum = 0
i = 1

while cardNumDigits > 0:
    digit = cardNumDigits % 10

    if i % 2 != 0:
        checkSum += digit
    else:
        digitDigit = digit * 2
        if digitDigit > 9:
            checkSum += digitDigit % 10
            checkSum += math.floor(digitDigit / 10)
        else:
            checkSum += digitDigit

    cardNumDigits = math.floor(cardNumDigits / 10)
    i += 1

if checkSum % 10 != 0:
    print("INVALID")
    sys.exit()

if l == 15:
    firstDigits = math.floor(cardNum / 10000000000000)
    if firstDigits == 34 or firstDigits == 37:
        print("AMEX")
        sys.exit()

if l == 16:
    firstDigits = math.floor(cardNum / 100000000000000)
    if firstDigits == 51 or firstDigits == 52 or firstDigits == 53 or firstDigits == 54 or firstDigits == 55:
        print("MASTERCARD")
        sys.exit()
    firstDigits = math.floor(cardNum / 1000000000000000)
    if firstDigits == 4:
        print("VISA")
        sys.exit()

if l == 13:
    firstDigits = math.floor(cardNum / 1000000000000)
    if firstDigits == 4:
        print("VISA")
        sys.exit()

print("INVALID")

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int get_long_len(long value);

int main(void)
{
    long cardNum = 0;
    do
    {
        cardNum = get_long("Number: ");
    }
    while (cardNum <= 0);

    int cardNumLen = get_long_len(cardNum);

    if (cardNumLen != 15 && cardNumLen != 16 && cardNumLen != 13)
    {
        printf("INVALID\n");
        return 0;
    }

    long cardNumDigits = cardNum;
    int checkSum = 0;

    for (int i = 1; cardNumDigits > 0; i++)
    {
        int digit = cardNumDigits % 10;

        if (i % 2 != 0)
        {
            checkSum += digit;
        }
        else
        {
            int digitDigit = digit * 2;
            if (digitDigit > 9)
            {
                checkSum += digitDigit % 10;
                checkSum += digitDigit /= 10;
            }
            else
            {
                checkSum += digitDigit;
            }
        }

        cardNumDigits /= 10;
    }

    if (checkSum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    if (cardNumLen == 15)
    {
        long firstDigits = cardNum / 10000000000000;

        if (firstDigits == 34 || firstDigits == 37)
        {
            printf("AMEX\n");
            return 0;
        }
    }

    if (cardNumLen == 16)
    {
        long firstDigits = cardNum / 100000000000000;

        if (firstDigits == 51 || firstDigits == 52 || firstDigits == 53 || firstDigits == 54 || firstDigits == 55)
        {
            printf("MASTERCARD\n");
            return 0;
        }

        firstDigits = cardNum / 1000000000000000;

        if (firstDigits == 4)
        {
            printf("VISA\n");
            return 0;
        }
    }

    if (cardNumLen == 13)
    {
        long firstDigits = cardNum / 1000000000000;

        if (firstDigits == 4)
        {
            printf("VISA\n");
            return 0;
        }
    }

    printf("INVALID\n");
    return 0;
}

int get_long_len(long value)
{
    int len = 1;
    while (value > 9)
    {
        len++;
        value /= 10;
    }
    return len;
}